### PR TITLE
MAILBOX-396 Some reactor improvement in Cassandra Mappers

### PR DESCRIPTION
**Pull request by chibenwa**
Avoid calling several time 'block' upon `add`, `setInMailboxes`, `findInMailbox`

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 1]
```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
...
+        return Mono.zip(messageUidMono, nextModSeqMono)
                 .doOnNext(tuple -> {
                     message.setUid(tuple.getT1());
                     message.setModSeq(tuple.getT2());
                 })
-                .block();
+                .then(Mono.just(message));
+    }
+
+    private <T> T block(Mono<T> mono) throws MailboxException {
```
Comment (Arsnael) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Is there really an advantage using this wrap method over a call to block() at the end of the chain?

Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: In order to preserve MailboxException handling, yes!

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 2]
```diff
@@ -171,17 +171,17 @@ public void delete(Mailbox mailbox, MailboxMessage message) {
...
+        return retrieveMessageIds(mailboxId, messageRange)
+            .flatMapMany(ids -> retrieveMessages(ids, ftype, Limit.from(max)))
             .map(MailboxMessage.class::cast)
             .sort(Comparator.comparing(MailboxMessage::getUid))
             .toIterable()
             .iterator();
     }
 
-    private List<ComposedMessageIdWithMetaData> retrieveMessageIds(CassandraId mailboxId, MessageRange messageRange) {
+    private Mono<List<ComposedMessageIdWithMetaData>> retrieveMessageIds(CassandraId mailboxId, MessageRange messageRange) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: why not `Flux<ComposedMessageIdWithMetaData>`

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 3]
```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
...
 
-        try {
-            Mono.zip(messageUidMono, nextModSeqMono)
+        return Mono.zip(messageUidMono, nextModSeqMono)
                 .doOnNext(tuple -> {
                     message.setUid(tuple.getT1());
                     message.setModSeq(tuple.getT2());
                 })
-                .block();
+                .then(Mono.just(message));
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: `thenReturn` ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 4]
```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
     public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        save(mailbox, addUidAndModseq(message, mailboxId))
-            .thenEmpty(indexTableHandler.updateIndexOnAdd(message, mailboxId))
-            .block();
+        block(addUidAndModseq(message, mailboxId)
+            .flatMap(Throwing.function(messageWithUidAndModSeq -> save(mailbox, messageWithUidAndModSeq)))
+            .thenEmpty(Mono.defer(() -> indexTableHandler.updateIndexOnAdd(message, mailboxId))));
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Why did you add this `defer`?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 5]
```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
     public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        save(mailbox, addUidAndModseq(message, mailboxId))
-            .thenEmpty(indexTableHandler.updateIndexOnAdd(message, mailboxId))
-            .block();
+        block(addUidAndModseq(message, mailboxId)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: should be replaced with an `onError*` method

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: List is needed down the line.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Overwize mutation on message didn't happen.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: not that easy, I'm not aware block can thorw checked exceptions.

Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Oh you are right

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: What about calling `window` then to have at the same time good latency and good throughput?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 4 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I don't understand, sorry. Could you elaborate?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Sounds awesome!

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: You are right, mutability is ugly, let's hide it.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 6]
```diff
@@ -384,10 +390,9 @@ public Flags getApplicableFlag(Mailbox mailbox) {
 
     private MessageMetaData setInMailbox(Mailbox mailbox, MailboxMessage message) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-
-        insertIds(addUidAndModseq(message, mailboxId), mailboxId)
-                .thenEmpty(indexTableHandler.updateIndexOnAdd(message, mailboxId))
-                .block();
+        block(addUidAndModseq(message, mailboxId)
+            .flatMap(messageWithUidAndModseq -> insertIds(messageWithUidAndModseq, mailboxId))
+            .thenEmpty(Mono.defer(() -> indexTableHandler.updateIndexOnAdd(message, mailboxId))));
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I still don't understand this `defer` sorry

**Comment by mbaechler** [comment]
Do we need load-testing on this one?

**Comment by chibenwa** [comment]
Maybe for find, a InboxHomeLoading could make sense

**Comment by chibenwa** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 7]
```diff
@@ -171,17 +171,18 @@ public void delete(Mailbox mailbox, MailboxMessage message) {
     @Override
     public Iterator<MailboxMessage> findInMailbox(Mailbox mailbox, MessageRange messageRange, FetchType ftype, int max) {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        return retrieveMessages(retrieveMessageIds(mailboxId, messageRange), ftype, Limit.from(max))
+        return retrieveMessageIds(mailboxId, messageRange)
+            .flatMap(ids -> retrieveMessages(ids, ftype, Limit.from(max)))
```
Comment (blackheaven) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: `flatMapSequential` instead of using `sort` may be better

Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: ids or not sorted when calling `retrieveMessageIds`

**Comment by Arsnael** [comment]
```
[b2a829507782ad728d139bf881a53175ed5067b8] Exception in thread "Thread-3059" java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:39805
[b2a829507782ad728d139bf881a53175ed5067b8] 	at org.apache.james.webadmin.WebAdminServer.lambda$start$1(WebAdminServer.java:95)
[b2a829507782ad728d139bf881a53175ed5067b8] 	at spark.Service.lambda$init$2(Service.java:582)
[b2a829507782ad728d139bf881a53175ed5067b8] 	at java.base/java.lang.Thread.run(Thread.java:834)
[b2a829507782ad728d139bf881a53175ed5067b8] Caused by: java.io.IOException: Failed to bind to /0.0.0.0:39805
```
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by chibenwa** [comment]
(build was aborted)

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[b2a829507782ad728d139bf881a53175ed5067b8] [ERROR] Failed to execute goal on project apache-james-backends-jpa: Could not resolve dependencies for project org.apache.james:apache-james-backends-jpa:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.checkerframework:checker-qual:jar:2.0.0: Failed to read artifact descriptor for org.checkerframework:checker-qual:jar:2.0.0: Could not transfer artifact org.checkerframework:checker-qual:pom:2.0.0 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.pom.part (No such file or directory) -> [Help 1]
```
test this please

**Comment by Arsnael** [comment]
```
[a04de8ca450a504772ba69debef5241d0bc10c51] [ERROR] Failures: 
[a04de8ca450a504772ba69debef5241d0bc10c51] [ERROR]   CassandraMailboxMapperTest.findMailboxWithPathLikeShouldRemoveDuplicatesAndKeepV2:919 
[a04de8ca450a504772ba69debef5241d0bc10c51] Expecting:
[a04de8ca450a504772ba69debef5241d0bc10c51]   <[Mailbox{namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=name}]>
[a04de8ca450a504772ba69debef5241d0bc10c51] to contain only:
[a04de8ca450a504772ba69debef5241d0bc10c51]   <[Mailbox{namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=name}]>
[a04de8ca450a504772ba69debef5241d0bc10c51] elements not found:
[a04de8ca450a504772ba69debef5241d0bc10c51]   <[Mailbox{namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=name}]>
[a04de8ca450a504772ba69debef5241d0bc10c51] and elements not expected:
[a04de8ca450a504772ba69debef5241d0bc10c51]   <[Mailbox{namespace=#private, user=Username{localPart=user, domainPart=Optional.empty}, name=name}]>
```
weird?

**Comment by chibenwa** [comment]
```
[221269b4c2eae0768c688236b150d1bd595a532f] [ERROR] Failed to execute goal on project javax-mail-extension: Could not resolve dependencies for project org.apache.james:javax-mail-extension:jar:3.5.0-SNAPSHOT: Could not transfer artifact org.checkerframework:checker-qual:jar:2.0.0 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.jar.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
```
[221269b4c2eae0768c688236b150d1bd595a532f] [ERROR] Failed to execute goal on project javax-mail-extension: Could not resolve dependencies for project org.apache.james:javax-mail-extension:jar:3.5.0-SNAPSHOT: Could not transfer artifact org.checkerframework:checker-qual:jar:2.0.0 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.jar.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
## Before

branch-master


## After

reactor

## Conclusion

While we see a stable p50 accross most commands we see a sharp drop of p99. (x10 globally)

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged
